src/main/java/org/broadinstitute/dsde/consent/ontology/datause/services/TextTranslationServiceImpl.java

```diff
@@ -89,13 +98,67 @@ public String translateDataset(String dataUseString) throws IOException {
         return translate(dataUse, TranslateFor.DATASET);
     }
 
+    @Override
+    public Map<String, Recommendation> translateParagraph(String paragraph) throws IOException {
+        return paragraph(paragraph);
+    }
+
     @Override
     public String translatePurpose(String dataUseString) throws IOException {
         ObjectMapper mapper = new ObjectMapper();
         DataUse dataUse = mapper.readValue(dataUseString, DataUse.class);
         return translate(dataUse, TranslateFor.PURPOSE);
     }
 
+    private Map<String, Recommendation> paragraph(final String paragraph) throws IOException {
+      Map<String, Recommendation> recommendations = new HashMap<>();
+
+      List<TermItem> terms = loadTermsFromGoogleStorage();
+
+      if (terms.isEmpty()) {
+        throw new IOException("Unable to process search terms");
+      }
+
+      for (TermItem term : terms) {
+        final String title = term.getTitle();
+        final String category = term.getCategory();
+        final String url = term.getUrl();
+        final String[] keywords = term.getKeywords();
+
+        for (String keyword : keywords) {
+          final boolean foundMatch = searchForKeyword(keyword, paragraph);
+          if (foundMatch) {
+            recommendations.computeIfAbsent(url, key -> new Recommendation(
+                title,
+                category
+            ));
+            break;
+          }
+        }
+      }
+
+      return recommendations;
+    }
+
+    private static boolean searchForKeyword(final String keyword, final String targetText) {
+      return StringUtils.containsIgnoreCase(targetText, keyword);
+    }
+
+    private List<TermItem> loadTermsFromGoogleStorage() {
+      try {
+        HttpResponse response = gcsStore.getStorageDocument("ontology/search-terms.json");
+        String terms = response.parseAsString();
+        return new Gson().fromJson(
+            terms,
+            new TypeToken<List<TermItem>>() {
+            }.getType()
```

Comment:
would this be constant? creating a new object here seems odd to me.

Reply:
`new TypeToken<List<TermItem>>() {}.getType()` is the type-specific way of getting a typed list class instead of just a generic list class. 

Here's another way to think of what this code is doing but doesn't work:
```
new Gson().fromJson(terms, List<TermItem>.class)
```